src/pages/files.tsx

```diff
@@ -1,64 +1,275 @@
+ * handles file uploads, and renders the file thumbnails.
+ */
+interface IFilesAppProps {
+  contentsManager: Contents.IManager;
+}
+
+function FilesApp(props: IFilesAppProps) {
+  const [listing, setListing] = useState<Contents.IModel | null>(null);
+  const [isUploading, setIsUploading] = useState(false);
+  const fileUploaderRef = useRef<IFileUploaderRef>(null);
+
+  const refreshListing = useCallback(async () => {
+    try {
+      const dirListing = await props.contentsManager.get('', { content: true });
+      setListing(dirListing);
+    } catch (err) {
+      await showErrorMessage(
+        'Error loading files',
+        `Could not load files from the contents manager: ${
+          err instanceof Error ? err.message : String(err)
+        }`
+      );
+    }
+  }, [props.contentsManager]);
+
+  useEffect(() => {
+    void refreshListing();
+  }, [refreshListing]);
+
   return (
-    <button className="je-Tile">
-      <div className="je-Tile-icon">
-        <props.icon.react />
+    <div className="je-FilesApp">
+      <FileUploader
+        ref={fileUploaderRef}
+        onUploadStart={() => setIsUploading(true)}
+        onUploadEnd={async () => {
+          setIsUploading(false);
+          await refreshListing();
+        }}
+        contentsManager={props.contentsManager}
+      />
+      <div className="je-FilesApp-content">
+        <div className="je-FilesApp-grid">
+          {/* "add new" tile */}
+          <div className="je-FileTile">
+            <div
+              className="je-FileTile-box"
+              onClick={() => fileUploaderRef.current?.triggerFileSelect()}
+            >
+              {isUploading ? (
+                <div className="je-FileTile-spinner" />
+              ) : (
+                <EverywhereIcons.addFile.react />
+              )}
+            </div>
+            <div className="je-FileTile-label">{isUploading ? 'Uploading file...' : 'add new'}</div>
+          </div>
+
+          {/* File thumbnails, and the rest of the tiles. */}
+          {listing &&
+            listing.type === 'directory' &&
+            (listing.content as Contents.IModel[])
+              .filter(f => {
+                return (
+                  f.type === 'file' &&
+                  isSupportedFileType({
+                    name: f.name,
+                    type: f.mimetype ?? '',
+                    size: f.size ?? 0,
+                    lastModified: Date.now()
+                  } as File)
+                );
+              })
+              .map(f => {
+                const fileIcon = getFileIcon(f.name, f.mimetype ?? '');
+                return (
+                  <div className="je-FileTile" key={f.path}>
+                    <div className="je-FileTile-box">
+                      <fileIcon.react />
+                    </div>
+                    <div className="je-FileTile-label">{f.name}</div>
+                  </div>
+                );
+              })}
+        </div>
```

Comment:
We use the same styles for the "add new" button and the file tiles, which is much simpler than the previous interface I had for recording uploaded files.